### PR TITLE
110598: expect() in semantic test producing unhelpful output

### LINKED ISSUE #110598: expect() in semantic test producing unhelpful output
Originally reported @ b/243693776

When writing a semantic test (that is expected to fail), I see an unhelpful error. Previously, the error showed the diff of the expected properties and the actual ones, but I believe an exception is being thrown while calculating this diff.

## Steps to Reproduce

1. Create a test using `matchesSemantics` or `containsSemantics` that is expected to fail
2. Run the test

**Expected results:** Test fails with error messaging that compares expected and actual output with no errors

**Actual results:** Test fails with error messaging containing errors

<details>
<summary>Code sample</summary>

```dart
      expect(
        tester.getSemantics(find.bySemanticsLabel('title')),
        matchesSemantics(
          hasTapAction: true,
          hasEnabledState: true,
          isEnabled: true,
          hasToggledState: true,
          isToggled: true,
          isFocusable: true,
          isSelected: true,
        ),
      );
```

</details>

<details>
  <summary>Logs</summary>

```
The following TestFailure was thrown running a test:
Expected: has semantics with actions:
[SemanticsAction:SemanticsAction.tap] without actions: [
            SemanticsAction:SemanticsAction.longPress,
            SemanticsAction:SemanticsAction.scrollLeft,
            SemanticsAction:SemanticsAction.scrollRight,
            SemanticsAction:SemanticsAction.scrollUp,
            SemanticsAction:SemanticsAction.scrollDown,
            SemanticsAction:SemanticsAction.increase,
            SemanticsAction:SemanticsAction.decrease,
            SemanticsAction:SemanticsAction.showOnScreen,
            SemanticsAction:SemanticsAction.moveCursorForwardByCharacter,
            SemanticsAction:SemanticsAction.moveCursorBackwardByCharacter,
            SemanticsAction:SemanticsAction.setSelection,
            SemanticsAction:SemanticsAction.copy,
            SemanticsAction:SemanticsAction.cut,
            SemanticsAction:SemanticsAction.paste,
            SemanticsAction:SemanticsAction.didGainAccessibilityFocus,
            SemanticsAction:SemanticsAction.didLoseAccessibilityFocus,
            SemanticsAction:SemanticsAction.dismiss,
            SemanticsAction:SemanticsAction.moveCursorForwardByWord,
            SemanticsAction:SemanticsAction.moveCursorBackwardByWord,
            SemanticsAction:SemanticsAction.setText
          ] with flags: [
            SemanticsFlag:SemanticsFlag.isButton,
            SemanticsFlag:SemanticsFlag.isFocusable,
            SemanticsFlag:SemanticsFlag.hasEnabledState,
            SemanticsFlag:SemanticsFlag.isEnabled,
            SemanticsFlag:SemanticsFlag.isInMutuallyExclusiveGroup
          ] without flags: [
            SemanticsFlag:SemanticsFlag.hasCheckedState,
            SemanticsFlag:SemanticsFlag.isChecked,
            SemanticsFlag:SemanticsFlag.isSelected,
            SemanticsFlag:SemanticsFlag.isSlider,
            SemanticsFlag:SemanticsFlag.isKeyboardKey,
            SemanticsFlag:SemanticsFlag.isLink,
            SemanticsFlag:SemanticsFlag.isTextField,
            SemanticsFlag:SemanticsFlag.isReadOnly,
            SemanticsFlag:SemanticsFlag.isFocused,
            SemanticsFlag:SemanticsFlag.isHeader,
            SemanticsFlag:SemanticsFlag.isObscured,
            SemanticsFlag:SemanticsFlag.isMultiline,
            SemanticsFlag:SemanticsFlag.namesRoute,
            SemanticsFlag:SemanticsFlag.scopesRoute,
            SemanticsFlag:SemanticsFlag.isHidden,
            SemanticsFlag:SemanticsFlag.isImage,
            SemanticsFlag:SemanticsFlag.isLiveRegion,
            SemanticsFlag:SemanticsFlag.hasToggledState,
            SemanticsFlag:SemanticsFlag.isToggled,
            SemanticsFlag:SemanticsFlag.hasImplicitScrolling
          ]
  Actual: SemanticsNode:<SemanticsNode#10(Rect.fromLTRB(0.0, 0.0,
103.0, 64.0), actions: [tap], flags: [isButton, hasEnabledState,
isEnabled, isFocusable], label: "title", textDirection:
ltr)>
   Which: null
'package:flutter/src/foundation/diagnostics.dart': Failed
assertion: line 3057 pos 5: 'indexOfDot != -1 && indexOfDot <
description.length - 1': The provided object "8" is not an enum.
at #0      _AssertionError._doThrowNew
(dart:core-patch/errors_patch.dart:51:61)
#1      _AssertionError._throwNew
(dart:core-patch/errors_patch.dart:40:5)
#2      describeEnum
(package:flutter/src/foundation/diagnostics.dart:3057:5)
#3      _MatchesSemanticsData.matches
(package:flutter_test/src/matchers.dart:2483:45)
#4      _expect (package:test_api/src/expect/expect.dart:123:30)
#5      expect (package:test_api/src/expect/expect.dart:46:3)
#6      expect
(package:flutter_test/src/widget_tester.dart:460:16)
#7      main.<anonymous closure>.<anonymous closure>
...
```

</details>

### PR DESCRIPTION
**Pull request by pdblasi-google**
Fixed issue with usage of `describeEnum` in `_MatchesSemanticsData` and added tests to prevent regressions.

Resolves #110598 

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter_test/test/matchers_test.dart [hunk 1]
```diff
@@ -733,6 +733,60 @@ void main() {
...
+          onTapHint: 'scan',
+          onLongPressHint: 'fill',
+          customActions: <CustomSemanticsAction>[
+            const CustomSemanticsAction(label: 'foo'),
+            const CustomSemanticsAction(label: 'bar'),
+          ],
+        ),
+      );
+
+      expect(failedExpectation, throwsA(isA<TestFailure>()));//.having((TestFailure failure) => failure., description, matcher)))
```
Comment (loic-sharma) [inline packages/flutter_test/test/matchers_test.dart:788]: I'm guessing the `.having...` comment is unintentional? 😄

Comment (pdblasi-google) [reply packages/flutter_test/test/matchers_test.dart:788]: It was. That plus analysis issues and some better messaging should be coming in shortly. 😛

Comment (pdblasi-google) [reply packages/flutter_test/test/matchers_test.dart]: Alright, should be good now! 😄

**Review by loic-sharma — APPROVED** [review]
LGTM but consider get a review from someone more familiar with this area

**Review by goderbauer — APPROVED** [review]
LGTM

**Pull request merged**